libs/utils/analysis_tool.py

```diff
@@ -0,0 +1,83 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copyright (C) 2017, ARM Limited, Google and contributors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import argparse
+import logging
+import os
+import select
+from trace import Trace
+from conf import LisaLogging
+from devlib.utils.misc import memoized
+
+class AnalysisTool(object):
+    """
+    A base class for LISA analysis command-line tools
+
+    This class intended to be subclassed in order to create a custom
+    analysis tool.
+    """
+
+    @memoized
+    def parse_args(self, parser):
+        if not parser:
+            parser = argparse.ArgumentParser(
+                    description='LISA Analysis tool configuration')
+
+        parser.add_argument('--trace', type=str,
+                default=None,
+                help='Path of the trace file or directory to be used',
+                required=True)
+
+        parser.add_argument('--log-stdout', action='store_true',
+                default=False,
+                help='Output logging to stdout',
+                required=False)
+
+        # Parse command line arguments
+        return parser.parse_args()
+
+    def run_analysis(self):
+	raise NotImplementedError('error')
+
+    def run_output(self):
+        if self.gui:
+            raise NotImplementedError('error')
+        print self.output
+
+    def __init__(self, parser=None, gui=False):
+	self.gui = gui
+        args = self.parse_args(parser)
+
+        self.args = args
+
+        # Analysis tools shouldn't output to standard output as their
+        # output can be used by other tools. Configure the logger to
+        # output only error messages
+        if not args.log_stdout:
+            LisaLogging.setup(level=logging.ERROR)
+        else:
+            LisaLogging.setup()
+
+        self._log = logging.getLogger('AnalysisTool')
+
+        # Create trace object
+        self._log.info('Creating trace object...')
+        self.trace = Trace(None, args.trace, [])
+
+	self.output = self.run_analysis()
+	self.run_output()
```

Comment:
This might just be my own weird obsession but I don't like when `__init__` does things other than up data structures - I'd much prefer to have everything ultimately driven by a `analysis_main` function or something, which is called from inside an `if __name__ == "__main__":` - it seems a lot more flexible.